File: baekjoon/tree_dp/2213.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const solution = (weights, input) => {
  const tree = input.reduce((prev, cur) => {
    if (prev.has(cur[0])) {
      prev.get(cur[0]).push(cur[1]);
    } else {
      prev.set(cur[0], [cur[1]]);
    }

    return prev;
  }, new Map());

  const dp = Array.from(new Array(weights.length + 1), () =>
    new Array(2).fill(0)
  );

  const max = (a, b) => (a > b ? a : b);
  const dfs = (idx) => {
    if (!tree.has(idx)) {
      dp[idx][0] = weights[idx - 1];
      dp[idx][1] = 0;
      return;
    }

    for (let child of tree.get(idx)) {
      dfs(child);
      dp[idx][0] += dp[child][1];
      dp[idx][1] += max(dp[child][0], dp[child][1]);
    }
    dp[idx][0] += weights[idx - 1];
  };

  const arr = [];
  const findArr = (idx, available) => {
    if (dp[idx][0] > dp[idx][1] && available) {
      arr.push(idx);
      available = false;
    } else {
      available = true;
    }
    if (!tree.has(idx)) return;
    for (let child of tree.get(idx)) {
      findArr(child, available);
    }
  };

  dfs(1);
  findArr(1, true);

  const max_size = max(dp[1][0], dp[1][1]);
  return [max_size, arr.sort()];
};

const input = [];
let weights = [];

let testcase = -1;

rl.on("line", function (line) {
  if (testcase === -1) {
    testcase = parseInt(line);
  } else if (testcase !== 0) {
    testcase = 0;
    weights = line.split(" ").map((v) => parseInt(v));
  } else {
    input.push(
      line
        .split(" ")
        .map((v) => parseInt(v))
        .sort()
    );
  }
}).on("close", function () {
  const [max_size, arr] = solution(weights, input);
  console.log(max_size);
  if (arr.length) console.log(arr.join(" "));

  process.exit();
});

File: baekjoon/tree_dp/2213.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.stream.Collectors;

public class Main {
    static Map<Integer, ArrayList<Integer>> board;
    static List<Integer> track;
    static int[] weights;
    static int[][] mem;
    static boolean[] visited;

    public static void dfs(int idx) {
        mem[idx][0] = 0;
        mem[idx][1] = weights[idx - 1];

        for (int child : board.get(idx)) {
            if (visited[child]) {
                continue;
            }
            visited[child] = true;
            dfs(child);
            mem[idx][0] += Math.max(mem[child][1], mem[child][0]);
            mem[idx][1] += mem[child][0];
        }
    }

    public static void trace(int idx, boolean available) {
        if (mem[idx][1] > mem[idx][0] && available) {
            track.add(idx);
            available = false;
        } else {
            available = true;
        }

        for (int child : board.get(idx)) {
            if (visited[child]) {
                continue;
            }
            visited[child] = true;
            trace(child, available);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        board = new HashMap<>();
        track = new ArrayList<>();

        int n = Integer.parseInt(br.readLine());
        weights = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        String line = "";
        while ((line = br.readLine()) != null && line.length() != 0) {
            int[] edge = Arrays.stream(line.split(" ")).mapToInt(Integer::parseInt).toArray();
            if (board.containsKey(edge[0])) {
                board.get(edge[0]).add(edge[1]);
            } else {
                board.put(edge[0], new ArrayList<>(Collections.singletonList(edge[1])));
            }
            if (board.containsKey(edge[1])) {
                board.get(edge[1]).add(edge[0]);
            } else {
                board.put(edge[1], new ArrayList<>(Collections.singletonList(edge[0])));
            }
        }

        visited = new boolean[n + 1];
        mem = new int[n + 1][2];
        visited[1] = true;
        dfs(1);
        int maxSize = Math.max(mem[1][0], mem[1][1]);

        visited = new boolean[n + 1];
        visited[1] = true;
        trace(1, true);
        System.out.println(maxSize);
        track.sort((a, b) -> a - b);
        System.out.println(track.stream().map(Object::toString).collect(Collectors.joining(" ")));
    }
}